torchtune/modules/loss/ppo.py

```diff
@@ -0,0 +1,98 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Optional, Tuple
+
+import torch
+import torch.nn as nn
+from torchtune.utils import ppo_utils
+
+
+class PPOLoss(nn.Module):
+    """
+    Proximal Policy Optimization (PPO) Loss module: https://arxiv.org/abs/1707.06347.
+
+    Using:
+        https://github.com/vwxyzjn/lm-human-preference-details/blob/ccc19538e817e98a60d3253242ac15e2a562cb49/lm_human_preference_details/train_policy_accelerate.py#L719
+        https://github.com/openai/baselines/blob/ea25b9e8b234e6ee1bca43083f8f3cf974143998/baselines/ppo2/model.py#L68-L75
+
+    as references.
+    Args:
+        gamma (float): Discount factor.
+        lmbda (float): lambda parameter for GAE.
+        epsilon (float): clipping parameter for PPO.
+    """
+
+    def __init__(
+        self,
+        gamma: float = 1.0,
+        lmbda: float = 0.95,
+        epsilon: float = 1e-5,
+        value_clip_range: float = 0.2,
+        value_coeff: float = 0.1,
+    ):
+        super().__init__()
+        self.gamma = gamma
+        self.lmbda = lmbda
+        self.epsilon = epsilon
+        self.value_clip_range = value_clip_range
+        self.value_coeff = value_coeff
+
+    def forward(
+        self,
+        pi_old_logprobs: torch.Tensor,
+        pi_logprobs: torch.Tensor,
+        advantages: torch.Tensor,
+        values: torch.Tensor,
+        returns: torch.Tensor,
+        padding_masks: Optional[torch.Tensor] = None,
+        value_padding_masks: Optional[torch.Tensor] = None,
+    ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        """
+        Forward pass of the PPO loss module.
+
+        Args:
+            pi_old_logprobs (torch.Tensor): Log probabilities of the old policy.
+            pi_logprobs (torch.Tensor): Log probabilities of the current policy.
+            advantage (torch.Tensor): Advantage values.
+            values (torch.Tensor): Value predictions.
+            returns (torch.Tensor): Return values.
+            padding_masks (Optional[torch.Tensor]): Padding token masks.
+            value_padding_masks (Optional[torch.Tensor]): Padding value masks.
+
+        Returns:
+            Tuple[torch.Tensor, torch.Tensor, torch.Tensor]: A tuple containing:
+                - loss: The total PPO loss.
+                - policy_loss: The policy function loss.
+                - value_loss: The value function loss.
+        """
+        ratio = torch.exp(pi_logprobs - pi_old_logprobs)
+        clipped_ratios = torch.clamp(ratio, 1.0 - self.epsilon, 1.0 + self.epsilon)
+
+        policy_losses = -advantages * clipped_ratios
+
+        # taking max instead of min to minimise loss
+        policy_loss = torch.max(policy_losses, -advantages * ratio)
```

Comment:
Shouldn't the second arg to torch.max be an int? (Again apologies if I'm missing something obvious here)

Reply:
great catch! i've always used torch.max for both (not sure if this is the right ref. to the internal code https://github.com/pytorch/pytorch/blob/217aac96d779841666527402fc113493b6bd6323/torch/onnx/symbolic_opset7.py#L38), but it looks like `torch.max` supports both. i'll change it to `torch.maximum` for clarity.